net/v2/parse.go

```diff
@@ -17,17 +17,59 @@ limitations under the License.
 package net
 
 import (
+	"fmt"
+	"net"
+
 	forkednet "k8s.io/utils/internal/third_party/forked/golang/net"
 )
 
-// ParseIPSloppy is identical to Go's standard net.ParseIP, except that it allows
-// leading '0' characters on numbers.  Go used to allow this and then changed
-// the behavior in 1.17.  We're choosing to keep it for compat with potential
-// stored values.
-var ParseIPSloppy = forkednet.ParseIP
-
-// ParseCIDRSloppy is identical to Go's standard net.ParseCIDR, except that it allows
-// leading '0' characters on numbers.  Go used to allow this and then changed
-// the behavior in 1.17.  We're choosing to keep it for compat with potential
-// stored values.
-var ParseCIDRSloppy = forkednet.ParseCIDR
+// ParseIP parses an IPv4 or IPv6 address to a net.IP. This accepts both fully-valid IP
+// addresses and irregular/ambiguous forms, making it usable for both validated and
+// non-validated input strings. It should be used instead of net.ParseIP (which rejects
+// some strings we need to accept for backward compatibility) and the old
+// netutilsv1.ParseIPSloppy.
+func ParseIP(ipStr string) (net.IP, error) {
+	// Note: if we want to get rid of forkednet, we should be able to use some
+	// invocation of regexp.ReplaceAllString to get rid of leading 0s in ipStr.
+	ip := forkednet.ParseIP(ipStr)
+	if ip != nil {
+		return ip, nil
+	}
+
+	if ipStr == "" {
+		return nil, fmt.Errorf("expected an IP address")
+	}
+	// NB: we use forkednet.ParseCIDR directly, not ParseIPNet, to avoid recursing
+	// between ParseIP and ParseIPNet.
+	if _, _, err := forkednet.ParseCIDR(ipStr); err == nil {
+		return nil, fmt.Errorf("expected an IP address, got a CIDR value")
+	}
+	return nil, fmt.Errorf("not a valid IP address")
+}
+
+// ParseIPNet parses an IPv4 or IPv6 CIDR string representing a subnet or mask, to a
+// *net.IPNet. This accepts both fully-valid CIDR values and irregular/ambiguous forms,
+// making it usable for both validated and non-validated input strings. It should be used
+// instead of net.ParseCIDR (which rejects some strings that we need to accept for
+// backward-compatibility) and the old netutilsv1.ParseCIDRSloppy.
+//
+// The return value is equivalent to the second return value from net.ParseCIDR. Note that
+// this means that if the CIDR string has bits set beyond the prefix length (e.g., the "5"
+// in "192.168.1.5/24"), those bits are simply discarded.
+func ParseIPNet(cidrStr string) (*net.IPNet, error) {
```

Comment:
Most ParseCIDRSloppy callers ignore the first argument anyway (https://github.com/danwinship/kubernetes/commit/dce0e89a48a09d23e90fb420667775763ded3182). Half of the ones that don't were trying to manually implement what `IPFromInterfaceAddr` does, so they can use that now. (The remainder will use ParseIPAsIPNet, coming in a few commits.)

And again, the consistent return value helps with generics.

Reply:
oh, the renaming from `ParseCIDR` to `ParseIPNet` is to be consistent with `ParseIP`, `ParseAddr`, and `ParsePrefix`; each function uses the basename of the type it returns.